### PR TITLE
Add AWS PrivateLink Controller

### PR DESCRIPTION
**Pull request by abhinavdahiya**
design for aws private link controller as recommended in https://github.com/openshift/hive/pull/1287

Details:

### create-cluster: supporting creating Internal clusters

Add `--internal` flag to the hiveutil create-cluster. When enabled
this configures the isntall-config.yaml with `publish: Internal`.
This is useful to create a cluster which does not expose the k8s API
and other cluster APIs to the Internet but only to the internal network.

### add awsprivatelink controller

The controller's goal is to setup a private link to the cluster being
created in user's AWS account such that Hive and other VPCs in Hub
account can reach the k8s API of the cluster using AWS Private Link and
not the Internet.

This allows even clusters that have k8s API published only to the VPC be
accessible to Hive using AWS Private Link.

A successful Private Link setup includes 4 resources,
1. VPC endpoint service pointing to the internal NLB of user's cluster
2. VPC endpoint connecting to the VPC endpoint service in Hub account
3. A Private Route53 Hosted zone in Hub account providing DNS resolution
  of the cluster's default API address `api.clustername.basedomain` to
  the VPC endpoint's regional DNS
4. Associations of the PHZ to Hive VPC(s) to allow DNS resolution setup
  in the previous step.

This assumes the network accessibility betweem Hive VPC(s) and the VPC
and the VPC where VPC endpoint is created is already setup.

To ensure that the controller created Enpoints in VPCs that are network
reachable, the controller consumes a globally configured list of VPCs
(inventory) that it should use.

```yaml
spec:
  awsPrivateLinkControllerConfig:
    vpcEndpointInventory:
    - region: (string)
      vpcID: (string)
      subnets:
      - id: (string)
        az: (string)
```

The controller uses the credentials configured in the ClusterDeployment
to create the Endpoint Service. To create resource in Hub account the
controller uses the globally configured credentials in HiveConfig.

```yaml
spec:
  awsPrivateLinkControllerConfig:
    credentialsSecretRef:
      name: (string)
```

The controller also needs to know which additional VPCs would require
the DNS resolution setup for the private link. This list should include
all the VPCs where various Hive clusters exist.

```yaml
spec:
  awsPrivateLinkControllerConfig:
    associatedVPCs:
    - id: (string)
      region: (string)
```

Reconciling VPC Endpoint service:

The controller discovers the internal NLB created by the installer. The
name of the NLB is `{infra_id}-int`.
The controller then lists for VPCEndpointServiceConfigurations using a
tag filter,

```
key: hive.openshift.io/private-link-access-for
value: {infra_id}
```

If no object exists with such tag, the controller creates a new
VPCEndpointServiceConfiguration using the discovered NLB.

The controller then reconciles the VPCEndpointServiceConfiguration to
make sure
- that Acceptance Required is False
- the NLBs used by the VPCEndpointServiceConfiguration is always only
  the cluster NLB
- the allows principals is always the AWS IAM entity defined by the
  hiveconfig.spec.awsPrivateLinkControllerConfig.credentialsSecretRef

Reconciling VPC Endpoint:

The controller list for VPCEndpoints using a tag filter,

```
key: hive.openshift.io/private-link-access-for
value: {infra_id}
```

If not object exists with such tag, the controller creates a new
Endpoint.

To create an Endpoint the controller first needs to figure out which VPC
to use. To choose a VPC the controller filters the inventory by these
steps
- filter out any inventory that doesn't match the region of the Endpoint
  Service
- filter out any inventory that doesn't have subnets in AZs that are
  supported by the Endpoint Service
- filter out any VPC that already has 255 VPC Endpoints created
The controller picks the first VPC that form the filtered down
inventory.

The controller then creates a VPC Endpoint in the VPC, subnet chosen
with privateDnsEnabled: False so that AWS does not create any PHZ for
the Endpoint.

Reconciling PHZ:

The controller gets the apiDomain from the admin kubeconfig secret, and
the endpoint from previous step.

The controller then uses the GetHostedZoneByVPC to find an existing Zone
in the VPC of the endpoint with name matching apiDomain. If no such
exists, the controller then creates a Private Route53 Hosted Zone with name
as the apiDomain, attaching it to the VPC of the Endpoint.

The controller the uses route53 UPSERT action to ensure that the PHZ has
a record pointing the apiDomain to the regional DNS name of the
endpoint.

Reconciling Associations of PHZ to desired VPCs:

THe controller then uses the list of desired associations from
HiveConfig and ensures that those VPCs are associated to the PHZ, also
removing any existing associations that are not in the desired list.

The controller updates the IDs of the resources in the
status.platformStatus.aws.privateLink section of the ClusterDeployment.

Failed Installations:
The controller async creates the private link resources for the
ClusterProvision attempt, and in case the provision fails when the
private link was already setup, the next ClusterProvision attempt
requires the private link resources to be deleted before it can proceed.
This is because it ClusterProvision attempt cleans up the previous
cluster resources before it tries again. Since the Endpoint Service
will block deletion of the cluster resources from previous failed
attempt, the controller must remove all the resources it created and
setup everthing again.

So the controller cleans up the resources using the prevInfraID on the
new ClusterProvision and the admin kubeconfig from the ClusterDeployment
from the previous attempt, and then requeues for another attempt of
createing the necessary resources.

the controller uses annotation
aws-private-link-controller.hive.openshift.io/last-cleanup-for and it's
value to track which infra_id clean has been already been done to decide
that this ClusterProvision can continue to be used for new set of
resources.

### operator: configure aws private link

The hive operator takes the spec.awsPrivateLinkControllerConfig and
stores it's json marshalled string into aws-private-link ConfigMap.
The operator rolls out a new version of hive-controller whenever the
contents change.

The hive-controller mounts the ConfigMap and exposes the path of the
config using the env AWSPrivateLinkControllerConfigFileEnvVar.

The aws private link controller is configured to read the configuration
from file specified by that env.

### manager: run awsprivatelink controller

Add the aws-private-link-controller to the list of controllers that are
run by the hive-controller's Manager

### create-cluster: allow enabling aws privatelink

Adds `--aws-private-link` flag to hiveutil create-cluster so that you
can generate a ClusterDeployment with private link enabled.
When set, it sets the spec.platform.aws.privateLink.enabled to True

### awsprivatelink: report progress and failure using conditions on cluster deployment

The controller uses `AWSPrivateLinkReady` and `AWSPrivateLinkFailed`
conditions on ClusterDeployment to provide progress update and failures
when setting up private link for a cluster.

setErrCondition

- sets the AWSPrivateLinkFailed condition with the reason and error
  message provided
- resets the AWSPrivateLinkReady condition with the same reason and
  error as the AWSPrivateLinkFailed condition

setProgressCondition

- reset the AWSPrivateLinkFailed condition when the progress is
  completed. This helps reducing the flapping of the failed condition
  when next reconcile progress from previous stages from where the failure
  happended.
- only updates the AWSPrivateLinkReady status when previous was False.
  This allows reconciles to prevent flapping of Ready once leveled by
  various reconcile steps.
- updates the reason and message of AWSPrivateLinkReady to match what
  was passed as argument.
  
### awsprivatelink: add unit test for various functions

Adds unit test for various functions and Reconcile scenarios.

### awsprivatelink: support associating VPCs across accounts

The Privte Hosted Zone for Endpoint is created in the same account as
the Endpoint using the credentials in
awsPrivateLinkControllerConfig.credentialsSecretRef .

But the VPCs that need to be associated to the PHZ for example the VPCs
where the Hive is running can each be in different VPCs. To associate
VPCs to PHZ in separate account, AWS requires additional steps detailed
in doc [1].

Therefore to support associating the VPCs from different accounts, the
controller needs credentials for account where the VPC lives.

```yaml
spec:
  awsPrivateLinkControllerConfig:
    associatedVPCs:
    - id: (string)
      region: (string)
      credentialsSecretRef: //optional
        name: (string)
```

So the controller uses the credentialsSecretRef defined for the VPC
items to associate the PHZ to the VPC whenever it is defined. If not
defined, the controller uses the credentials used for creating VPC
endpoints etc to attach the VPC.

So when VPC are in separate account, the steps become
- AuthorizeAssociation of VPC to PHZ using
  awsPrivateLinkControllerConfig.credentialsSecretRef
- Associate VPC to PHZ using associatedVPCs[$idx].credentialsSecretRef
- RemoveAuthorizeAssociation of VPC to PHZ using
  awsPrivateLinkControllerConfig.credentialsSecretRef

[1]: https://docs.aws.amazon.com/Route53/latest/DeveloperGuide/hosted-zone-private-associate-vpcs-different-accounts.html

xref: https://issues.redhat.com/browse/HIVE-1407

### CODE REVIEW

File: config/crds/hive.openshift.io_hiveconfigs.yaml
Comment (gregsheremeta) [inline config/crds/hive.openshift.io_hiveconfigs.yaml:51]: s/Privat/Private/

File: pkg/apis/hive/v1/validating-webhooks/clusterdeployment_validating_admission_hook.go
Comment (twiest) [inline pkg/apis/hive/v1/validating-webhooks/clusterdeployment_validating_admission_hook.go:541]: Was this on purpose?

If so, why not just remove the whole statement?

File: pkg/constants/constants.go
Comment (twiest) [inline pkg/constants/constants.go:325]: Hmm. AFAIK we've never done a config file for a controller. I'm fine with it, but want to make sure that this is a pattern we now want to adopt for other controllers as well.

@dgoodwin thoughts?  ^^

**Review by dgoodwin — CHANGES_REQUESTED** [review]
I haven't made it all the way through yet but here's a good chunk, maybe half way so far, but need to take a break.

One thing that I think we're going to need is docs/aws-private-link.md, unfortunately as I know that's going to be a big ask, but I think QE, SD, and even myself as a reviewer could really use it. I'd love the docs to encompass AWS setup, HiveCOnfig setup, and ClusterDeployment use, and a quick section on how it works adapted from your design doc.

File: pkg/controller/awsprivatelink/cleanup.go
Comment (dgoodwin) [inline pkg/controller/awsprivatelink/cleanup.go:100]: cleanupVPCEndpoint

File: pkg/controller/awsprivatelink/cleanup.go
Comment (dgoodwin) [inline pkg/controller/awsprivatelink/cleanup.go:101]: Endpoint

File: pkg/controller/awsprivatelink/awsprivatelink_controller.go
Comment (dgoodwin) [inline pkg/controller/awsprivatelink/awsprivatelink_controller.go:459]: could not

File: pkg/controller/awsprivatelink/awsprivatelink_controller.go
Comment (abhinavdahiya) [reply pkg/controller/awsprivatelink/awsprivatelink_controller.go:206]: Conditions tend to be APIs, but reasons are usually just one word summarizes for categorizations. So i don't think these should be part of the API.

File: pkg/controller/awsprivatelink/awsprivatelink_controller.go
Comment (abhinavdahiya) [reply pkg/controller/awsprivatelink/awsprivatelink_controller.go:220]: i think the intent is clear this way, that we need to do work but later on.

File: pkg/controller/awsprivatelink/awsprivatelink_controller.go
Comment (abhinavdahiya) [reply pkg/controller/awsprivatelink/awsprivatelink_controller.go:266]: i think the intent is clear this way, that we need to do work but later on.

**Review by abhinavdahiya — COMMENTED** [review]
pushed fixes for
- validation webhook
- godoc and most spelling fixes

File: apis/hive/v1/hiveconfig_types.go
Comment (abhinavdahiya) [reply apis/hive/v1/hiveconfig_types.go]: if these are requires, the more appropriate place is controller's package docs. but since there are no other packages that use that in the repo.. i'll defer to the use/configure doc that is already required.

File: pkg/constants/constants.go
Comment (dgoodwin) [reply pkg/constants/constants.go:329]: Yes definitely outside of this change.

**Review by abhinavdahiya — COMMENTED** [review]
i think i have pushed changes for all the requests and replied to the questions and comments.
So i think this is ready for next round @dgoodwin

File: docs/awsprivatelink.md
Comment (abhinavdahiya) [reply docs/awsprivatelink.md:45]: > Are these Endpoint VPCs on public internet?

VPCs are always private networks. additional resources have to be created to make them access internet or have internet accessible services in the VPC.

File: docs/awsprivatelink.md
Comment (abhinavdahiya) [reply docs/awsprivatelink.md:59]: it's an inventory for creating VPC endpoints, while the items themselves are VPCs. So personally the current naming sounds more appropriate to me.  🤷🏽 if you want it to be `envpointVPCInventory` ?

File: docs/awsprivatelink.md
Comment (abhinavdahiya) [reply docs/awsprivatelink.md:68]: you mean using an actual AWS region for example.

File: docs/awsprivatelink.md
Comment (abhinavdahiya) [reply docs/awsprivatelink.md:85]: these are not just hive VPCs, but any VPC that will need to access the cluster api like they would have if the clsuter was on the internet. this can and should include any jump network that SREs might use to access the cluster API.

File: docs/awsprivatelink.md
Comment (abhinavdahiya) [reply docs/awsprivatelink.md:120]: > Out of curiosity, how do you manually test? Have you been testing all within our AWS account just using multiple VPCs or is a second separate account required?

my testing has been in the same VPC as the Hive. cross accounts requires networking setup that is too much for manual testing.

File: docs/awsprivatelink.md
Comment (dgoodwin) [reply docs/awsprivatelink.md:68]: More so what the subnets should look like.

Comment (abhinavdahiya) [reply docs/awsprivatelink.md:68]: > More so what the subnets should look like.

we don't care what the subnets should look like.. creating VPC and then subnets in VPC is a fairly basic AWS networking task. the only requirement we have subnet in all AZs in the region. rest is upto the admins.

we know the limit of vpc endpoints in VPC is 255, so i can include that make sure there are atleast 255 usable addresses in each subnet.. other than that, i don't understand what is appropriate here.

File: docs/awsprivatelink.md
Comment (dgoodwin) [reply docs/awsprivatelink.md:85]: Sounds good.

File: docs/awsprivatelink.md
Comment (dgoodwin) [reply docs/awsprivatelink.md:59]: Yeah something seems off here. I'm going to switch over to commenting on the hiveconfig_types.go

File: apis/hive/v1/hiveconfig_types.go
Comment (dgoodwin) [inline apis/hive/v1/hiveconfig_types.go:114]: So here we have vpcEndpointInventory but below we have associatedVPCs (inverts the position of VPC). Both look like a list of VPC IDs and regions, one with additional subnet array.

One is using VPCID, the other just ID.

Was the AWSPrivateLinkVPC meant to be embedded in AWSPrivateLinkInventory as well? Right now it's separated but only used in AWSAssociatedVPC.

Feels like these are very similar and should look quite similar in the API.

endpointVPCInventory
- region
  vpcID
  subnets
associatedVPCs:
- region
  vpcID

File: pkg/controller/awsprivatelink/awsprivatelink_controller.go
Comment (abhinavdahiya) [reply pkg/controller/awsprivatelink/awsprivatelink_controller.go:212]: @dgoodwin added similar shouldSync for the controller in 062a2ce

File: apis/hive/v1/hiveconfig_types.go
Comment (abhinavdahiya) [reply apis/hive/v1/hiveconfig_types.go]: @dgoodwin  fixed the api for hiveconfig in f3d3954 

new one looks like:
```yaml
endpointsVPCInventory:
- region
  vpcID:
  subnets:
  - availabilityZone:
    subnetID:
associatedVPCs:
- region:
  vpcID:
```

also include the doc updates for real world example for subnets and network ranges

File: pkg/controller/awsprivatelink/awsprivatelink_controller.go
Comment (dgoodwin) [inline pkg/controller/awsprivatelink/awsprivatelink_controller.go:405]: One thing I'm still worried about here, if there's a problem mid-way and we end up re-reconciling, there are going to be a lot of writes to ClusterDeployment status. We'll fail, set the failed condition, and then requeue, which will move the progress condition back through each state it already completed. It still feels like we need to track the state we're in here, or perhaps just monitor "did we do anything" a little better, and then only update the condition if we found something has to be done. In this case that would mean we drop the above set progress, and in the block below, if we can't find the NLB, we set WaitingForNLB reason.

File: pkg/controller/awsprivatelink/awsprivatelink_controller.go
Comment (dgoodwin) [inline pkg/controller/awsprivatelink/awsprivatelink_controller.go:438]: Similar here, if we drop this set progress, and instead in the block below we track if we had to do anything or not as a change bool from reconcileVPCEndpointService, and if it's true we set Ready progress to ReconciledVPCEndpointService. I think essentially I'm proposing the condition reasons change from "what i'm about to do" to " what I last did", this will cut down on the etcd writes in the event of requeues, and seemingly going back in time to steps we previously did successfully if you're watching the condition.

Granted if we start getting failures it won't be as obvious where in the Ready condition which would now show us the last thing we did but not the thing we're doing currently, but it will be visible in the Failure condition, so in tandem it should be quite clear where the issue is surfacing.

Same for each Ready progress update below.

**Review by dgoodwin — COMMENTED** [review]
Just a couple typo requests still pending, and one last issue I was thinking about last night, then I think this is ready to go.

File: pkg/controller/awsprivatelink/awsprivatelink_controller.go
Comment (abhinavdahiya) [reply pkg/controller/awsprivatelink/awsprivatelink_controller.go]: i'm hoping 
43e3e78  is what you had in mind @dgoodwin

Comment (dgoodwin) [reply pkg/controller/awsprivatelink/awsprivatelink_controller.go]: Yes this should do the trick, thanks.

You have some conflicts, otherwise LGTM, squash anything you want to.

**Comment by abhinavdahiya** [comment]
@dgoodwin rebased to fix the conflicts and squashed all controllers changes to one commit.

**Pull request merged**
